### ISSUE TITLE
Use map instead of batch submit in local.get_async

### ISSUE THREAD

**Issue by SebastienDorgan** [dask/dask#8334]
The _local.get_async_ function only uses the submit function even when the actual scheduler used is a _concurrent.futures.Executor_ and _chunksize > 1_ or a call to the _map_ method would be more efficient. Using the _map_ method would also allow another slightly different approach to be implemented where chunking is delegated to the executor. That is, all tasks could be sent simultaneously using _map_ and the optimization of task distribution would be delegated to the provided executor.

**Comment by SebastienDorgan** [comment]
Indeed I missed this part of the code! Thanks for your quick answer

**Comment by jakirkham** [comment]
Yeah we did look at using `map`. Unfortunately as it doesn't return a `Future`, but instead returns a blocking iterator, it is kind of hard to work with in the context of the Dask scheduler. Ideally there would be something like `map_async`, which would return a `Future` (for [example]( https://stackoverflow.com/a/35909364 )). Though Idk if that kind of thing is in-scope for the `Executor` interface (as it may involve more machinery under-the-hood).

The batching functionality with `chunksize` should allow one to batch execute multiple tasks in a single job. Though as you note this is a different thing.

There are probably other improvements that can be made to the local scheduler. Of particular interest is a first class `Future` object ( https://github.com/dask/dask/issues/6380 ) to allow things like `persist` to work nicely on the local scheduler (many use cases for this listed in that issue). This may require other changes to occur in how the local scheduler behaves.
